io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,258 @@
+        let l_bname_ptr = base_dev_name.as_ptr() as *mut i8;
+        let l_cname_ptr = cache_dev_name.as_ptr() as *mut i8;
+
+        let (s, r) = oneshot::channel::<ErrnoResult<()>>();
+
+        let spdk_ftl_conf_size = mem::size_of::<spdk_ftl_conf>() as u64;
+        let mut ftl_conf = spdk_ftl_conf {
+            ..unsafe { mem::zeroed() }
+        };
+        unsafe { spdk_ftl_get_default_conf(&mut ftl_conf, spdk_ftl_conf_size) };
+        ftl_conf.name = l_name_ptr;
+        ftl_conf.base_bdev = l_bname_ptr;
+        ftl_conf.cache_bdev = l_cname_ptr;
+        ftl_conf.fast_shutdown = true;
+        ftl_conf.verbose_mode = true;
+        // SPDK_FTL_MODE_CREATE = (1 << 0)
+        ftl_conf.mode = 1 << 0;
+
+        let errno = unsafe {
+            bdev_ftl_create_bdev(
+                &mut ftl_conf,
+                Some(ftl_bdev_init_fn_cb),
+                cb_arg(s),
+            )
+        };
+
+        if errno != 0 {
+            let err = BdevError::CreateBdevFailed {
+                source: Errno::from_raw(errno.abs()),
+                name: self.name.clone(),
+            };
+
+            error!("{:?} error: {}", self, err.verbose());
+
+            return Err(err);
+        }
+
+        r.await
+            .context(bdev_api::BdevCommandCanceled {
+                name: self.name.clone(),
+            })?
+            .context(bdev_api::CreateBdevFailed {
+                name: self.name.clone(),
+            })?;
+
+        if let Some(mut bdev) = UntypedBdev::lookup_by_name(&self.name) {
+            if let Some(uuid) = self.uuid {
+                unsafe { bdev.set_raw_uuid(uuid.into()) };
+            }
+
+            if !bdev.add_alias(&self.alias) {
+                warn!("{:?}: failed to add alias '{}'", self, self.alias);
+            }
+
+            return Ok(self.get_name());
+        }
+
+        Err(BdevError::BdevNotFound {
+            name: self.get_name(),
+        })
+    }
+
+    async fn destroy(self: Box<Self>) -> Result<(), Self::Error> {
+        debug!("{:?}: deleting", self);
+
+        let mut result =
+            if let Some(mut bdev) = UntypedBdev::lookup_by_name(&self.name) {
+                bdev.remove_alias(&self.alias);
+                let (s, r) = oneshot::channel::<ErrnoResult<()>>();
+
+                unsafe {
+                    bdev_ftl_delete_bdev(
+                        (*bdev.unsafe_inner_ptr()).name,
+                        true,
+                        Some(done_errno_cb),
+                        cb_arg(s),
+                    );
+                }
+
+                r.await
+                    .context(bdev_api::BdevCommandCanceled {
+                        name: self.name.clone(),
+                    })?
+                    .context(bdev_api::DestroyBdevFailed { name: self.name })
+            } else {
+                Err(BdevError::BdevNotFound { name: self.name })
+            };
+        if let Err(e) = bdev_destroy(&self.bbdev_uri).await {
```

Comment:
Shouldn't we reverse the order here?
Destroy the backing bdevs first, and then the self bdev?

Reply:
`bdev_ftl_delete_bdev` requires the backing bdev's to be present.

Reply:
nvm I must have been day dreaming, makes perfect sense to destroy the top-down...